src/compiler/parser.ts

```diff
@@ -2207,39 +2242,37 @@ module ts {
+            // FormalParameters[Yield,GeneratorParameter,Await,AsyncParameter] : (Modified)
             //      ...
             //
-            // FormalParameter[Yield,GeneratorParameter] :
-            //      BindingElement[?Yield, ?GeneratorParameter]
+            // FormalParameter[Yield,GeneratorParameter,Await,AsyncParameter] : (Modified)
+            //      BindingElement[?Yield,?GeneratorParameter,?Await,?AsyncParameter]
             //
-            // BindingElement[Yield, GeneratorParameter ] : See 13.2.3
-            //      SingleNameBinding[?Yield, ?GeneratorParameter]
-            //      [+GeneratorParameter]BindingPattern[?Yield, GeneratorParameter]Initializer[In]opt
-            //      [~GeneratorParameter]BindingPattern[?Yield]Initializer[In, ?Yield]opt
+            // BindingElement[Yield,GeneratorParameter,Await,AsyncParameter] : (Modified) See 13.2.3
+            //      SingleNameBinding[?Yield,?GeneratorParameter,?Await,?AsyncParameter]
+            //      [+GeneratorParameter]BindingPattern[?Yield,?Await,GeneratorParameter]Initializer[In]opt
+            //      [+AsyncParameter]BindingPattern[?Yield,?Await,GeneratorParameter]Initializer[In]opt
+            //      [~GeneratorParameter,~AsyncParameter]BindingPattern[?Yield,?Await]Initializer[In,?Yield,?Await]opt
             //
-            // SingleNameBinding[Yield, GeneratorParameter] : See 13.2.3
+            // SingleNameBinding[Yield,GeneratorParameter,Await,AsyncParameter] : (Modified) See 13.2.3
             //      [+GeneratorParameter]BindingIdentifier[Yield]Initializer[In]opt
-            //      [~GeneratorParameter]BindingIdentifier[?Yield]Initializer[In, ?Yield]opt
+            //      [+AsyncParameter]BindingIdentifier[Await]Initializer[In]opt
+            //      [~GeneratorParameter,~AsyncParameter]BindingIdentifier[?Yield,?Await]Initializer[In,?Yield,?Await]opt
             if (parseExpected(SyntaxKind.OpenParenToken)) {
-                var savedYieldContext = inYieldContext();
-                var savedGeneratorParameterContext = inGeneratorParameterContext();
-
-                setYieldContext(yieldAndGeneratorParameterContext);
-                setGeneratorParameterContext(yieldAndGeneratorParameterContext);
-
-                var result = parseDelimitedList(ParsingContext.Parameters, parseParameter);
-
-                setYieldContext(savedYieldContext);
-                setGeneratorParameterContext(savedGeneratorParameterContext);
-
+                var explicitFlags = ParserContextFlags.None;
+                if (yieldAndGeneratorParameterContext) {
+                    explicitFlags |= ParserContextFlags.Yield | ParserContextFlags.GeneratorParameter;
+                }
+                if (awaitAndAsyncParameterContext) {
+                    explicitFlags |= ParserContextFlags.Await | ParserContextFlags.AsyncParameter;
+                }
+                var result = setContextParametersAnd(parseDelimitedParameterList, explicitFlags);

```

Comment:
i think i'd prefer AmbientFlags be non optinoal, and have this code explicitly pass along Flags.None here.